### PR TITLE
Add agx scheduler

### PR DESCRIPTION
**Pull request by obrask1**
Added AGX scheduler with the following logic:

- 1.) Start listening for virtual vehicle connections by defining and opening a socket 
- 2.) accept all readable client sockets and put them in a queue. Listen continuously until timeout is reach (customizable, currently at 1 second)
- 3.) If there is at least 1 job in queue, execute the first job in the queue
- 4.) When training is completed, remove the job from the queue and close the connection with that virtual vehicle
- 5.) repeat from 2)

### CODE REVIEW

File: communication/client/client.py [hunk 1]
```diff
@@ -1,55 +1,85 @@
...
+                )
+                print("[CLIENT] Terminating server session...")
                 break
 
-        self.s.close()
-        print("[CLIENT] Socket closed.")
+        self.socket_server.close()
+        print("[CLIENT] Server socket closed.\n")
 
     def _client_do_task(self, server_message):
```
Comment (jonasandersson10) [inline communication/client/client.py:50]:
```suggestion
    def _train(self, server_message):
```
maybe we could rename this method now to something more specific?

File: communication/scheduler/scheduler.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            # use select to determine readable sockets
+            readable_sockets, _, _ = select.select([self.socket], [], [], self.TIMEOUT)
+
+            if readable_sockets:
+                for readable_socket in readable_sockets:
+                    client_socket, client_address = readable_socket.accept()
+
+                    # add the client socket to the list of connected virtual vehicles
+                    self.BUFFER.append((client_socket, client_address))
+                    print(f"[AGX] Adding client {client_address} to que")
```
Comment (jonasandersson10) [inline communication/scheduler/scheduler.py:53]:
```suggestion
                    print(f"[AGX] Adding client {client_address} to queue")
```

File: communication/scheduler/scheduler.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            print("[AGX] Invalid messaged received from vv.")
+            print(
+                f"[AGX] Received message: {virtual_vehicle_message}. Ignoring message."
+            )
+            stop_session = True
+
+        return stop_session
+
+    def _train(self):
+        time.sleep(4)
```
Comment (jonasandersson10) [inline communication/scheduler/scheduler.py:116]:
```suggestion
        time.sleep(4)
        results = {'weights': []}
        return results
```

File: communication/scheduler/scheduler.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        stop_session = False
+
+        if virtual_vehicle_message == self.message_from_virtual_vehicle[0]:
+            print(f"[AGX] Received msg from virtual vehicle: {virtual_vehicle_message}")
+            conn.sendall(self.message_to_virtual_vehicle[0].encode("utf-8"))
+
+        elif virtual_vehicle_message == self.message_from_virtual_vehicle[1]:
+            print(f"[AGX] Received msg from virtual vehicle: {virtual_vehicle_message}")
+            print("[AGX] Training started")
+            self._train()
```
Comment (jonasandersson10) [inline communication/scheduler/scheduler.py:100]:
```suggestion
            results = self._train()
```

File: communication/scheduler/scheduler.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            print("[AGX] Training completed")
+            print(f"[AGX] Ending session with vehicle {addr}.")
+            conn.sendall(self.message_to_virtual_vehicle[1].encode("utf-8"))
```
Comment (jonasandersson10) [inline communication/scheduler/scheduler.py:103]: maybe we can try to send back a `pickle` - version of the results? Should imitate a more realistic scenario

**Review by jonasandersson10 — COMMENTED** [review]
Really great work @obrask1 💪 ! Nice integration of the AGX.

The only suggestions I have is to push the communication to something even more realistic, such as a pickled dictionary of weights. In my comments below, I suggested to implement that between VV<->AGX but maybe we'd also like to have it between AZServer <-> VV?

File: communication/scheduler/scheduler.py (hunk 5 above)
Comment (obrask1) [reply communication/scheduler/scheduler.py:103]: agree! Just like an empty dict as below or something more?

**Review by obrask1 — COMMENTED** [review]
good comments. All implemented in new commit.

File: communication/scheduler/scheduler.py (hunk 5 above)
Comment (jonasandersson10) [reply communication/scheduler/scheduler.py:103]: yes I think that'll do for now when we are just setting up the framework! 👍

File: communication/client/client.py [hunk 6]
```diff
@@ -1,55 +1,85 @@
...
+        self.socket_server.send(self.message_to_server[0].encode("utf-8"))
 
         while True:
 
-            data = self.s.recv(self.MSG_LENGTH)
+            data = self.socket_server.recv(self.MSG_LENGTH)
 
-            if str(data, "utf-8") == "TRAIN":
+            if str(data, "utf-8") == self.message_from_server[0]:
                 self._client_do_task(str(data, "utf-8"))
```
Comment (jonasandersson10) [inline communication/client/client.py:38]:
```suggestion
                self._train(str(data, "utf-8"))
```

**Review by jonasandersson10 — COMMENTED** [review]
found a bug. also, I get 
`Error response from daemon: Ports are not available: exposing port TCP 0.0.0.0:59999 -> 0.0.0.0:0: listen tcp 0.0.0.0:59999: bind: address already in use`
when I try to deploy with docker-compose. do you get the same?

**Comment by jonasandersson10** [comment]
> @obrask1 found a tiny bug.
> 
> Also, I get the following error when I try to deploy with docker-compose. Do you get the same?
> 
> `Error response from daemon: Ports are not available: exposing port TCP 0.0.0.0:59999 -> 0.0.0.0:0: listen tcp 0.0.0.0:59999: bind: address already in use`

Ok, it doesn't seem to be a problem with the deployment itself, rather on my computer. Just changing the port number made it work.

File: conda-environment-files/conda-environment-cpu-unix.yml [hunk 7]
```diff
@@ -37,15 +37,16 @@ dependencies:
   - opencv-python
   - dropbox==11.36.0
   - typer~=0.7.0
-  - pyproj==3
+  # - pyproj==3
```
Comment (jonasandersson10) [inline conda-environment-files/conda-environment-cpu-unix.yml:40]: Do we want to comment this back on? Or maybe remove the pkg?

File: conda-environment-files/conda-environment-cpu-unix.yml [hunk 8]
```diff
@@ -37,15 +37,16 @@ dependencies:
...
+  - pyproj
   - plotly==5.11.0
   - ipykernel>=5
   - zod[cli]==0.1.8
   - imageio
   - timezonefinder
   - flwr[simulation]
   - pyyaml
-  - psycopg2
+  # - psycopg2
```
Comment (jonasandersson10) [inline conda-environment-files/conda-environment-cpu-unix.yml:49]: shall we remove this perhaps?

**Review by jonasandersson10 — COMMENTED** [review]
Really awesome work @obrask1! 🚀  Just two final questions on the commented pkgs. Otherwise it looks great!

File: conda-environment-files/conda-environment-cpu-unix.yml (hunk 8 above)
Comment (obrask1) [reply conda-environment-files/conda-environment-cpu-unix.yml]: yeah it was not intentional to add these... reverted in new commit

File: conda-environment-files/conda-environment-cpu-unix.yml (hunk 7 above)
Comment (obrask1) [reply conda-environment-files/conda-environment-cpu-unix.yml]: yeah it was not intentional to add these... reverted in new commit

**Review by jonasandersson10 — APPROVED** [review]
Great work! 🚀

**Pull request merged**